Stop blacklisting Go 1.22+, drop Go < 1.21 support, use Go 1.22 in CI

kolyshkin opened the pull request:
High level overview:
 - remove nsenter compile restriction since Go 1.22.4 and Go 1.23 can now be used. THIS PART NEEDS TO BE BACKPORTED TO 1.1.
 - switch to Go 1.22 in CI whether possible (everywhere except Ubuntu 20.04 and CentOS 7 and 8).
 - set minimum requirement to Go 1.21
 - rm some old code.

See individual commits for details.

rata commented on README.md:
```diff
@@ -29,6 +29,11 @@ A third party security audit was performed by Cure53, you can see the full repor
+NOTE there is a known issue of runc's nsenter package being incompatible
+with systems having glibc < 2.32 when compiled with Go 1.22 older than 1.22.4.
+For those systems, runc refuses to compile. The workaround is to use Go 1.21
+or Go 1.23.
```
maybe like this?
```suggestion
, Go 1.23 or Go >= 1.22.4.
```

kolyshkin replied:
So, Go 1.22.4 would work, but the issue is, I can't disable the kludge in `libcontainer/nsenter/nsenter_go122.go` for Go >= 1.22.4, there's just no way to do it. I mention this problem in the first commit as well in this PR description.

I mean, we could add some kind of a hack to Makefile which checks go version and disables this...

kolyshkin replied:
OK, ended up adding a hack to Makefile to check for go1.22 version, and removed the README addition (since if runc can't be compiled it prints a self-explanatory error).

PTAL

rata reviewed (approved):
LGTM, thanks!

It took me a while to test locally that the go version dance was fine, but it indeed seem fine :)

AkihiroSuda commented on Makefile:
```diff
@@ -23,6 +23,12 @@ LDFLAGS_COMMON := -X main.gitCommit=$(COMMIT) -X main.version=$(VERSION)
 GOARCH := $(shell $(GO) env GOARCH)
 
+# Allow Go 1.22.4+ to compile nsenter. This is ugly.
+GO122PATCH := $(shell $(GGO) version | sed -n 's/.* go1\.22.\([0-9]*\) .*/\1/p')
```
`GGO` seems a typo

kolyshkin replied:
Thanks! fixed.

kolyshkin commented:
We'll probably need to port the first 3 patches to release-1.1 (so that it will work with Go 1.22).

kolyshkin commented:
> btw, it seems the ubuntu builds are stalled.

This PR drops go 1.20.x tests, but they are marked as required in repo config -- this is why they are shown as "Expected -- waiting ...". I will fix the repo config accordingly (rm go 1.20.x, require go 1.22.x) once this PR will be ready to be merged (I guess once Go 1.21.4 is out).

kolyshkin commented:
@opencontainers/runc-maintainers wdyt?

AkihiroSuda commented on .github/workflows/test.yml:
```diff
@@ -24,16 +24,20 @@ jobs:
         dmz: ["", "runc_nodmz"]
         exclude:
+          # Go 1.22 + glibc < 2.32 + nsenter does not play well (issue 4233),
+          # thus nsenter package won't compile. Disable ubuntu 20.04 + go 1.22.
```
Maybe add "TODO: enable this again when Go 1.22.4 is available"

kolyshkin replied:
done, thanks!

AkihiroSuda commented on README.md:
```diff
@@ -27,7 +27,7 @@ A third party security audit was performed by Cure53, you can see the full repor
 ## Building
 
-`runc` only supports Linux. It must be built with Go version 1.19 or higher.
+`runc` only supports Linux. It must be built with Go version 1.21 or higher.
```
Maybe add some note about Go 1.22

rata commented on Makefile:
```diff
@@ -23,6 +23,12 @@ LDFLAGS_COMMON := -X main.gitCommit=$(COMMIT) -X main.version=$(VERSION)
+# Allow Go 1.22.4+ to compile nsenter. This is ugly.
+GO122PATCH := $(shell $(GO) version | sed -n 's/.* go1\.22.\([0-9]*\) .*/\1/p')
+ifeq ($(shell test $(GO122PATCH) -ge 4 && echo yes), yes)
+	export CGO_CPPFLAGS+="-D GO1_22_WORKS=1"
```
Super nit, we are not using c++, so CFLAGS seem more appropriate (go passes both to C programs, though)

```suggestion
	export CGO_CFLAGS+="-D GO1_22_WORKS=1"
```

kolyshkin replied:
I guess you're mixing up CPPFLAGS (for C PreProcessor) and CXXFLAGS (C++).

rata replied:
Oh, yes, sorry. `.cpp` being used as file extension for c++ projects really confuses me with the CPPFLAGS :-D

rata reviewed (approved):
LGTM

kolyshkin commented:
@lifubang @cyphar PTAL

AkihiroSuda commented:
CentOS tests seem failing

thaJeztah commented:
> CentOS tests seem failing

For CentOS stream 8, that may be expected; it's EOL since May 31 (just got hit by that in some CI pipeline that als had it in the Matrix)

AkihiroSuda commented:
> > CentOS tests seem failing
> 
> For CentOS stream 8, that may be expected; it's EOL since May 31 (just got hit by that in some CI pipeline that als had it in the Matrix)

I guess we now have to replace CentOS with Alma, Rocky, or Oracle Linux?

lifubang commented:
This one should be merged after go 1.22.4 has released.

kolyshkin commented:
Go 1.22.4 is out, this PR is rebased with the following change:

```diff
diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 4c40b47b..7155c163 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -30,11 +30,6 @@ jobs:
         criu: ["", "criu-dev"]
         dmz: ["", "runc_nodmz"]
         exclude:
-          # Go 1.22 + glibc < 2.32 + nsenter does not play well (issue 4233),
-          # thus nsenter package won't compile. Disable ubuntu 20.04 + go 1.22.
-          # TODO: remove this once Go 1.22.4 is out.
-          - os: ubuntu-20.04
-            go-version: 1.22.x
           # Disable most of criu-dev jobs, as they are expensive
           # (need to compile criu) and don't add much value/coverage.
           - criu: criu-dev
```

rata reviewed (approved):
LGTM.

The ubuntu 20.04 build is failing, probably 1.22.4 didn't make it on that action (the error is the C error about the version)

kolyshkin commented:
Ah, `actions/setup-go` still installs go 1.22.3 (given go 1.22.x). They always lag behind (see https://github.com/actions/setup-go/issues/407).

Have to wait a few days I guess (or specify the version number explicitly, which I don't want to).

kolyshkin commented:
@lifubang PTAL (except for pinning go in ci to 1.22.4 this is all good, and we can revert this patch later)

AkihiroSuda commented:
Needs rebase

AkihiroSuda commented on Makefile:
```diff
@@ -23,6 +23,12 @@ LDFLAGS_COMMON := -X main.gitCommit=$(COMMIT) -X main.version=$(VERSION)
+GO122PATCH := $(shell $(GO) version | sed -n 's/.* go1\.22.\([0-9]*\) .*/\1/p')
+ifeq ($(shell test $(GO122PATCH) -ge 4 && echo yes), yes)
+	export CGO_CPPFLAGS+="-D GO1_22_WORKS=1"
+endif
```
I guess we can now just block Go 1.22.0-1.22.3 ?

kolyshkin replied:
You mean, block it from the top-level Makefile, and remove nsenter_go122 entirely? Makes sense

AkihiroSuda replied:
Yes

lifubang replied:
But for some libct users may need this file?

lifubang replied:
How about write some code to support all go 1.22 versions?

lifubang replied:
Please see #4310, feel free to close it if we don't want to support go 1.22.3, I think if we can support all go versions, it will be better.

AkihiroSuda replied:
Let's just skip 1.22.3, for simplicity

lifubang replied:
But for libct users, such as hashicorp/nomad, if they don’t write such method in Makefile, they will get an error even in go 1.22.4.

kolyshkin replied:
OTOH we are checking glibc version in nsenter_go122 and it's somewhat hard to do from the Makefile.

kolyshkin replied:
Leaving the current hack in place, added some extra comments.

kolyshkin replied:
> if we don't want to support go 1.22.3, I think if we can support all go versions, it will be better.

To me it doesn't make much sense to support those older versions. We can just require Go 1.22 >= 1.22.4, it's a decent requirement.

> But for libct users, such as hashicorp/nomad, if they don’t write such method in Makefile, they will get an error even in go 1.22.4.

That's right. 

Ideally, I'd like to do it all in one place -- that is, `nsenter_go122.go`. Unfortunately, I can't find a way to determine go's patch version from there.

kolyshkin replied:
> I can't find a way to determine go's patch version from there.

... during compile time.

rata replied:
Is it so bad to not handle the case at all, then? I mean, not do anything for go 1.22, just document in the function/package that if using go 1.22, you need at least go 1.22.4 (or glibc and explain it all) and be done with it.

Go 1.22.4 also includes security fixes, so hopefully people upgrade faster.

A runc release will at least take 1-2 more weeks I think, so go 1.22.4 should be out for a few weeks already by then.

rata replied:
IMHO, I'd leave this PR without handling the 1.22 case, we can discuss what to do for 1.22 #4310

lifubang replied:
I agree, because go 1.22.4 has released, we think that people should swith to go 1.22.4, they will not use go 1.22.3, especially when running CI.

kolyshkin replied:
Yeah, maybe we should just stop blacklisting go and instead add a note that for Go 1.22 one should use Go 1.22.4+.

rata reviewed (approved):
LGTM

kolyshkin commented on README.md:
```diff
@@ -29,6 +29,10 @@ A third party security audit was performed by Cure53, you can see the full repor
 
 `runc` only supports Linux. It must be built with Go version 1.19 or higher.
 
+NOTE: if building with Go 1.21.x, make sure to use 1.21.4 or a later version
```
thanks! I need more coffee this morning :)

kolyshkin commented:
Assuming that by the time 1.2.0 is out, everyone will use Go 1.22.x  >= 1.22.4, I just dropped the restriction (and added a note to README).

I think this is ready to be merged. WDYT @cyphar?

lifubang commented on README.md:
```diff
@@ -29,6 +29,10 @@ A third party security audit was performed by Cure53, you can see the full repor
 
 `runc` only supports Linux. It must be built with Go version 1.19 or higher.
```
In line 30, I have tested it and found that runc can’t compile in go 1.19 anymore, we should change it to ‘go 1.20 or higher’.

lifubang replied:
Note: This is an error review if we’re doing review commit by commit.
